### ISSUE TITLE
Tested the blob !

### ISSUE THREAD

**Issue by delahee** [Tiago-Ling/JitterDemo#1]
Hi, I tested the things, some remarks :

1- impossible to compile with a regular3.0.1 haxe, I had to butcher your Updatable interface.
2- bloating : you use an interface over bitmap, this is not necessary and provokes dynamic access which are both slow and blurr you code space.
3- openfl namespace : don't use it, prefer flash namespace because when you have to test without openfl, it very convenient and allow you to bisect problems comming from openfl form those comming from flash.
4- int statuses can be uplifted to enum
5- I cannot launch anything with my 3.0.1 conf (nothing appears) ... if you could lift the sample to a regular flash sample, it could be helpful

As a general rule, you might want to expose a more simplified example :)

gl, hope it helps !

++

ps : managed to launch it in neko, i ll push further...

**Comment by delahee** [comment]
finally got it to run on neko and flash and windows.

Ok several problems about rendering:
- you use 1 pixel grid pattern, the jitter is normal then as you don't snap on integer coordinates 
- the backdrop is displayed as is, if you put everything on a light backdrop, less stuttter as the gpu will not have to sample hard contrasts
- gc triggering sometimes cause double draw

Other than that, when i fix the pattern and the backdrop, the rendering is correct, i don't see any problem aside the usual LCD latency problems which a good patternised background will fix.

There are a lot of gc stutter but that is more dued to neko+openfl internals which allocates too much.
As a general fact, fast scrolling is very hard to achieve with GC'ed engines because you don't control the allocations and then the framerate is not smooth.

it seems there are badly detached object when swithing modes (which openfl does not report) it can hint about a potential multiple child problem.

**Comment by Tiago-Ling** [comment]
Hi David, thanks for taking the time to test this :). Some questions regarding your fixes:
- What do you mean by `1 pixel grid pattern`? 
- What do you mean by `light backdrop`?

As for the GC, i don't know if anything can be done - in this case there is no GC triggering in my code right? 

By the way i'm gonna push some changes to the code to make it simpler as you suggested. We are also testing scrolling in a very simple test using pure Lime and the issue is still there. Talking to Hugh and Joshua it seems this can be some problem with SDL (when using Windows target, which is the one i'm really interested in).

Thanks again for your great help!

**Comment by delahee** [comment]
1- the windows of the test (girders?) are not well suited for scrolling because they are too thin
2- put the game on a lighter color 
As for the gc, you should test in flash with scout enabled, this way you'll know what happpens.

As i said, there is not one single issue ;), I do not believe the issue lies in the windows target.

So you adress better the issue, test in flash and compare with windows to be sure about what is the problem.

**Comment by Tiago-Ling** [comment]
Hi @delahee,

I greatly simplified all tests for all libraries (also added tests in pure Lime and NME), changing the image, background color and removing all unnecessary code.

Basically all the tests do now is:
- Instantiate an image on the white screen (the logo for each framework);
- Move the image from right to left, resetting its position when it goes off screen;
- Use this delta time logic for changing the object position (listening to the `ENTER_FRAME` event):

``` haxe
function update(e:Event) {
    var ticks:Float = haxe.Timer.stamp() * 1000.0;
    var delta:Float = (ticks - lastTicks);

    bitmap.x -= (delta / 1000) * speed;
    if (bitmap.x + bitmap.width < 0)
        bitmap.x = stage.stageWidth;

    lastTicks = ticks;
}
```

Also, now for the `BitmapDemo` class i'm using only the flash package, with no mentions to OpenFL whatsoever, this hopefully will make it easier for you to test between AS3 and cpp targets.

At least here, all demos show the problem, Flash among them, apparently with even more than the others. I don't know if i can make these simpler than they are now, maybe changing the delta logic (passing a fixed hardcoded value or something).

As far as the discussion is going in OpenFL's Slack channel, people are thinking this may be a problem at cpp level (using SDL as the native backend for Lime and NME).

All machines i have run these so far present the jittering - if you can confirm the occurrence in yours it'll be a great help.

Please let me know if you have more suggestions to try and isolate this nasty problem :)

Thanks a lot!

**Comment by delahee** [comment]
I see three things still that would be issues in a retail game
- you still have to snap pixel pos to integer values
- The reset code is not ok, it should read

If ( x > max)
X -= max

Or
X = x%max;
if x <0 x +=max

Or some looping thing like that because full reset without catching up the
difference will cause problems too
- Remove and add from the display list is not necessary and can cause
  allocations, prefer a sprite.visible = true/false if parent stays the same.

I ll test when I can.

Maybe it's a problem with sdl but you can still learn tricks from this lore
;)

**Comment by Tiago-Ling** [comment]
I'm gonna update the code to allow pixel snapping, but in my experience (the tests had a flag to allow the user to toggle snap to pixels before) this _increases the jitter_ instead. By pixel snapping i take that you mean this (or some variation of this):

``` haxe
bitmap.x -= Math.ffloor((delta / 1000) * speed);
```

The other two points i did not understand / are not applicable to the current code (?):
- There is no remove and add to the display list during the tests. The only place where i remove something that was added before is in the `Menu` class in which i remove the TextFields that show the keys to access each test - there is no messing with the display list in the tests, mere repositioning of the image during enter frame;
- I agree that the reset code could be better for a seamless scrolling background implementation like in the initial tests (making the two moving images appear to be an infinite one). That is not the case anymore as there is only one image which gets reset to its original position only when it is fully outside the screen - there is no positioning gap in this operation and this does not influence the jitter.

All tests do the exact same thing, changing only the render and / or the update loop method but if you can only test one of them please take a look at the `BitmapDemo.hx` file, it is less than 50 lines now and does not depend on any other user-created classes like before (also using only the flash package as you suggested). I really appreciate the help!

:)

**Comment by delahee** [comment]
For snapping I use

Update (...)
Bmp.x = Std.int( Ent.x)

That way we never accumulate error that += can wield.

As for the reset, I ll check out what you did, it makes me think that maybe
the sprite simply goes too fast and do simply no longer get rendered enough
so that eye persistance can no longer help movement interpolation.
You might want to scale the whole scene a magnitude bigger and see if there
is still a problem.

I ll check out current version with flash output and compare with windows
target.

As a trick to test these kind of issue, I used to film the render and
examinate it frame by frame.

**Comment by Tiago-Ling** [comment]
Good idea about the filming - i suppose you have to record without any compression and codecs to get all frames?

Also, when you go test it, make sure to also try `30 fps`, this will make the problem much more evident.

**Comment by azrafe7** [comment]
I tested the demo too, hoping to find a solution for it... Here's what I found so far:

It seems that what contributes the most to the appearance of the issue is the extreme variability of FPS. I've slightly modified the openfl FPS counter to print the istantaneous FPS, instead of an average (https://gist.github.com/azrafe7/7f36a82d0fee7b97ce53). Which showed that, depending on the target and the expected FPS, the measured one varies highly from one frame to the next one.

On my machine also flash often struggles to keep a fixed 30fps (while it settles nicely on 60fps when told to do so). It's interesting that on my box (win7), targeting cpp & neko for 70fps, also results in a smoother animation, without the FPS seesaw I get at 60fps.

Googling around it seems that there's no universal solution to fix this thing (and it's indeed a common one), although you can try to alleviate the problem with interpolation, fixed timestep and other tricks, the general opinion (from what I gather) is that avoiding GC runs, enable vsync and trying to mask the ugly effect with other in-game objects (like reducing contrast or adding more "action") is the best you can do.

Some of the links I've read through:
- http://stackoverflow.com/questions/1284886/optimizing-transition-movement-smoothness-for-a-2d-flash-game
- http://www.java-gaming.org/topics/solving-stuttering-with-fixed-timesteps-once-and-for-all/24975/view.html
- http://answers.unity3d.com/questions/228095/why-is-the-motion-jerky-in-a-simple-2d-game.html
